include/deal.II/base/standard_tensors.h

```diff
@@ -0,0 +1,513 @@
+   * @f[
+   *  \textrm{trace}(\textrm{Dev} \{ \bullet \}) = 0 \, .
+   * @f]
+   *
+   * @note Comparing the definition of this tensor in Holzapfel (2007) to that
+   * adopted here (and used in Wriggers (2008)), the inclusion of the extra
+   * factor $J^{-2/dim}$ does not, at the outset, seem to be a reasonable choice.
+   * However, in the author's view it makes direct implementation of the
+   * expressions for isochoric (referential) stress contributions and their
+   * linearisation simpler in practise.
+   */
+  template <typename Number>
+  static SymmetricTensor<4, dim, Number>
+  Dev_P (const Tensor<2, dim, Number> &F);
+
+  /**
+   * Returns the transpose of the fourth-order referential deviatoric tensor,
+   * as constructed from the deformation gradient tensor @p F.
+   * The result performs the following operation:
+   * @f[
+   *  \textrm{Dev}^{T} \{ \bullet \} = J^{-2/\textrm{dim}}[ \{ \bullet \} -
+   *  [1/\textrm{dim}][\mathbf{C}^{-1} : \{ \bullet\}] \mathbf{C} ]
+   *  := \mathcal{P}^{T} : \{ \bullet\}
+   * @f]
+   */
+  template <typename Number>
+  static SymmetricTensor<4, dim, Number>
+  Dev_P_T (const Tensor<2, dim, Number> &F);
+
+  /**
+   * Returns the negative derivative of the inverse of the right Cauchy-Green
+   * tensor with respect to the right Cauchy-Green tensor itself,
+   * as constructed from the deformation gradient tensor @p F.
+   * The result, aligned with Wriggers (2008) and accounting for symmetry, is
+   * defined in index notation as
+   * @f[
+   *  -\mathcal{E}_{abcd} := -\frac{1}{2}[ C^{-1}_{ac}C^{-1}_{bd}
+   *    + C^{-1}_{ad}C^{-1}_{bc}  ]
+   * @f]
+   */
+  template <typename Number>
+  static SymmetricTensor<4, dim, Number>
+  EE (const Tensor<2, dim, Number> &F);
+};
+
+}
+
+
+
+// ------------------------- inline functions ------------------------
+
+
+
+template <int dim>
+const SymmetricTensor<2, dim>
+StandardTensors::Elasticity<dim>::I = unit_symmetric_tensor<dim>();
+
+
+
+template <int dim>
+template <typename Number>
+inline
+Tensor<2, dim, Number>
+StandardTensors::Elasticity<dim>::F (const Tensor<2, dim, Number> &Grad_u)
+{
+ // See Wriggers (2008) p 23 equ. 3.14
+ return I + Grad_u;
+}
+
+
+
+template <int dim>
+template <typename Number>
+inline
+SymmetricTensor<2, dim, Number>
+StandardTensors::Elasticity<dim>::C (const Tensor<2, dim, Number> &F)
+{
+ // See Wriggers (2008) p 23 equ. 3.15 (text below)
+ return symmetrize(transpose(F)*F);
+}
+
+
+
+template <int dim>
+template <typename Number>
+inline
+SymmetricTensor<2, dim, Number>
+StandardTensors::Elasticity<dim>::b (const Tensor<2, dim, Number> &F)
+{
+ // See Wriggers (2008) p 23 equ. 3.25 (text below)
+ return symmetrize(F*transpose(F));
+}
+
+
+
+template <int dim>
+template <typename Number>
+inline
+SymmetricTensor<2, dim, Number>
+StandardTensors::Elasticity<dim>::E (const Tensor<2, dim, Number> &F)
+{
+ // See Wriggers (2008) p 23 equ. 3.15
+ return Number(0.5)*(C(F) - unit_symmetric_tensor<dim,Number>());
```

Comment:
Would it work if we made sure that `operator-` and `operator+` do what they are expected to do? I would expect that we could declare them as something like
```
  template <int dim, typename Number1, typename Number2>
  SymmetricTensor<2,dim,decltype(Number1() - Number2())>
  operator- (....) {....}
```
This requires C++11, but would at least do the right thing.

I'm just wondering where the pain is best spent. Yes, we could write things in such a way that we track the template argument through everything. Or we could spend the time overloading a few operators. I don't know what's better, but I've learned that having template arguments on everything is often no fun either.